python/ray/data/_internal/compute.py

```diff
@@ -203,73 +203,79 @@ def map_block_nosplit(
+                    block, meta = blocks_in.pop()
+                    if context.block_splitting_enabled:
+                        ref = worker.map_block_split.remote(block, meta.input_files)
+                    else:
+                        ref, meta_ref = worker.map_block_nosplit.remote(
+                            block, meta.input_files
+                        )
+                        metadata_mapping[ref] = meta_ref
+                    tasks[ref] = worker
+                    block_indices[ref] = len(blocks_in)
+                    tasks_in_flight[worker] += 1
+
+            map_bar.close()
+            new_blocks, new_metadata = [], []
+            # Put blocks in input order.
+            results.sort(key=block_indices.get)
+            if context.block_splitting_enabled:
+                for result in ray.get(results):
+                    for block, metadata in result:
+                        new_blocks.append(block)
+                        new_metadata.append(metadata)
             else:
-                ready_workers.add(worker)
-                map_bar.set_description(
-                    "Map Progress ({} actors {} pending)".format(
-                        len(ready_workers), len(workers) - len(ready_workers)
-                    )
-                )
-
-            # Schedule a new task.
-            while (
-                blocks_in
-                and tasks_in_flight[worker] < self.max_tasks_in_flight_per_actor
-            ):
-                block, meta = blocks_in.pop()
-                if context.block_splitting_enabled:
-                    ref = worker.map_block_split.remote(block, meta.input_files)
-                else:
-                    ref, meta_ref = worker.map_block_nosplit.remote(
-                        block, meta.input_files
-                    )
-                    metadata_mapping[ref] = meta_ref
-                tasks[ref] = worker
-                block_indices[ref] = len(blocks_in)
-                tasks_in_flight[worker] += 1
-
-        map_bar.close()
-        new_blocks, new_metadata = [], []
-        # Put blocks in input order.
-        results.sort(key=block_indices.get)
-        if context.block_splitting_enabled:
-            for result in ray.get(results):
-                for block, metadata in result:
+                for block in results:
                     new_blocks.append(block)
-                    new_metadata.append(metadata)
-        else:
-            for block in results:
-                new_blocks.append(block)
-                new_metadata.append(metadata_mapping[block])
-            new_metadata = ray.get(new_metadata)
-        return BlockList(new_blocks, new_metadata)
+                    new_metadata.append(metadata_mapping[block])
+                new_metadata = ray.get(new_metadata)
+            return BlockList(new_blocks, new_metadata)
+
+        except Exception:
+            for worker in workers:
+                ray.kill(worker)
+            raise
```

Comment:
Could we

```
catch Exception as e:
  try:
     kill workers...
  except Exception:
     logger.exception("Error killin..")
  finally:
      raise e
 ```
 
 To preserve the original exception?

Reply:
Absolutely - doing this now!